datasync/chaindatafetcher/kafka/kafka.go

```diff
@@ -0,0 +1,164 @@
+	once   sync.Once
+)
+
+type KafkaBroker struct {
+	producer   sarama.AsyncProducer
+	admin      sarama.ClusterAdmin
+	brokers    []string
+	handlers   map[string]func(*sarama.ConsumerMessage) error
+	consumer   *Consumer
+	replicas   int16
+	partitions int32
+}
+
+func New(groupID string, brokerList []string, replicas int16, partitions int32) types.EventBroker {
+	once.Do(func() {
+		kb = &KafkaBroker{
+			brokers:    brokerList,
+			handlers:   map[string]func(*sarama.ConsumerMessage) error{},
+			replicas:   replicas,
+			partitions: partitions,
+		}
+		kb.newClusterAdmin()
+		kb.newProducer()
+
+		// TODO-ChainDataFetcher context has to be passed by outside.
+		kb.consumer = NewConsumer(context.Background(), kb.newConsumer(groupID))
+	})
+
+	return kb
+}
+
+func (r *KafkaBroker) Publish(topic string, msg interface{}) error {
+	r.CreateTopic(topic)
+	item := &sarama.ProducerMessage{
+		Topic: topic,
+		Key:   sarama.StringEncoder(topic),
+	}
+	if v, ok := msg.(types.IKey); ok {
+		item.Key = sarama.StringEncoder(v.Key())
+	}
+	data, err := json.Marshal(msg)
+	if err != nil {
+		return err
+	}
+	item.Value = sarama.StringEncoder(data)
+
+	r.producer.Input() <- item
+
+	return nil
+}
+
+func (r *KafkaBroker) Subscribe(topic string, handler interface{}) error {
+	r.CreateTopic(topic)
+	h, ok := handler.(func(*sarama.ConsumerMessage) error)
+	if !ok {
+		return errors.New("unsupported type")
+	}
+
+	return r.consumer.Subscribe(topic, h)
+}
+
+func (r *KafkaBroker) CreateTopic(topic string) (types.Topic, error) {
+	err := r.admin.CreateTopic(topic, &sarama.TopicDetail{
+		NumPartitions:     r.partitions,
+		ReplicationFactor: r.replicas,
+	}, false)
+
+	return types.Topic{Name: topic}, err
+}
+
+func (r *KafkaBroker) DeleteTopic(topic string) error {
+	return r.admin.DeleteTopic(topic)
+}
+
+func (r *KafkaBroker) ListTopics() ([]types.Topic, error) {
+	topics, err := r.admin.ListTopics()
+	if err != nil {
+		return nil, err
+	}
+
+	var ret []types.Topic
+	for k := range topics {
+		ret = append(ret, types.Topic{Name: k})
+	}
+
+	return ret, nil
+}
+
+func (r *KafkaBroker) newProducer() {
+	config := sarama.NewConfig()
+	config.Producer.RequiredAcks = sarama.WaitForLocal
+	config.Producer.Compression = sarama.CompressionSnappy
+	config.Producer.Flush.Frequency = 500 * time.Millisecond
+
+	producer, err := sarama.NewAsyncProducer(r.brokers, config)
+	if err != nil {
+		logger.Crit("Failed to start Sarama producer", "err", err)
+	}
+
+	r.producer = producer
+}
+
+func (r *KafkaBroker) newConsumer(groupID string) sarama.ConsumerGroup {
+	config := sarama.NewConfig()
+	config.Version = sarama.MaxVersion
+	config.Consumer.Group.Session.Timeout = 6 * time.Second
+	config.Consumer.Group.Heartbeat.Interval = 2 * time.Second
+
+	id, _ := uuid.GenerateUUID()
+	config.ClientID = fmt.Sprintf("%s-%s", groupID, id)
+
+	consumer, err := sarama.NewConsumerGroup(r.brokers, groupID, config)
+	if err != nil {
+		logger.Crit("NewConsumerGroup is failed", "err", err)
```

Comment:
How about leaving configurations in the log?

Reply:
Added on 1f989f2dcfbb1c9eeed58c88629d79559f246f26.